integration/hibernate-base/src/main/java/com/blazebit/persistence/integration/hibernate/base/HibernateJpaProvider.java

```diff
@@ -845,4 +847,43 @@ public boolean needsTypeConstraintForColumnSharing() {
     public void setCacheable(Query query) {
         query.setHint("org.hibernate.cacheable", true);
     }
+
+    @Override
+    public List<String> getIdentifierOrUniqueKeyEmbeddedPropertyNames(EntityType<?> owner, String attributeName) {
+        // Hibernate does not properly optimize the access to the natural id but instead always creates a join for associations
+        boolean supportsNaturalIdAccessOptimization = false;
+        AbstractEntityPersister entityPersister = getEntityPersister(owner);
+        Type propertyType = entityPersister.getPropertyType(attributeName);
+        org.hibernate.type.EntityType entityType = (org.hibernate.type.EntityType) propertyType;
+        List<String> identifierOrUniqueKeyPropertyNames;
+        Type identifierOrUniqueKeyType = entityType.getIdentifierOrUniqueKeyType(entityPersister.getFactory());
+
+        if (identifierOrUniqueKeyType instanceof EmbeddedComponentType) {
+            EmbeddedComponentType embeddedComponentType = (EmbeddedComponentType) identifierOrUniqueKeyType;
+            identifierOrUniqueKeyPropertyNames = Arrays.asList(embeddedComponentType.getPropertyNames());
+        } else {
+            String identifierOrUniqueKeyPropertyName = entityType.getIdentifierOrUniqueKeyPropertyName(entityPersister.getFactory());
+            identifierOrUniqueKeyPropertyNames = Collections.singletonList(identifierOrUniqueKeyPropertyName);
+        }
+
+        if (supportsNaturalIdAccessOptimization) {
+            return identifierOrUniqueKeyPropertyNames;
+        }
+
+        EntityPersister attributeEntityPersister = entityPersisters.get(entityType.getAssociatedEntityName());
+        Type identifierType = attributeEntityPersister.getIdentifierType();
+
+        if (identifierType instanceof EmbeddedComponentType) {
+            EmbeddedComponentType embeddedComponentType = (EmbeddedComponentType) identifierType;
+            if (!identifierOrUniqueKeyPropertyNames.containsAll(Arrays.asList(embeddedComponentType.getPropertyNames()))) {
+                return Collections.emptyList();
+            }
+        } else {
+            if (!identifierOrUniqueKeyPropertyNames.contains(attributeEntityPersister.getIdentifierPropertyName())) {
+                return Collections.emptyList();
```

Comment:
Wasn't this supported in Hibernate?

Reply:
The idea of this code is to return an empty list for associations that are joined by natural id to prevent the id access optimization from kicking in